### PR TITLE
Update toolchain to Clang 17

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/7193, close https://github.com/pingcap/tiflash/issues/8890

Problem Summary:

### What is changed and how it works?

Use clang-17.0.6 for CI/CD building since this PR merged. Other branches keep using clang-13.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

tpch50, 1 tidb, 1 tikv, 1 tiflash, https://github.com/pingcap/tiflash/tree/c15fcc2bf83c6e1230c43a0daaccdc00fb15304f
```
Query  clang13 (s)   clang17 (s)
Q1      1.71            1.71
Q2      0.85            0.83
Q3      1.51            1.47
Q4      5.39            4.46
Q5      2.87            2.79
Q6      0.44            0.43
Q7      1.59            1.58
Q8      2.24            2.21
Q9      8.23            7.95
Q10     1.91            1.87
Q11     0.57            0.62
Q12     1.02            1.07
Q13     2.09            2.13
Q14     0.57            0.57
Q15     1.21            1.23
Q16     0.50            0.53
Q17     4.33            4.41
Q18     3.87            3.91
Q19     0.93            0.84
Q20     0.89            0.85
Q21     3.55            3.46
Q22     0.35            0.34
Total  49.06           48.28
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by purelind** [comment]
Hi, because here specify image_tag_suffix in .toolchain.yml, to ensure that build-common can pull images properly, need to confirm if this image `hub.pingcap.net/tiflash/tiflash-llvm-base:amd64-llvm-14.0.6` exists, if not, need to build the image first and upload it.

ci groovy file https://github.com/PingCAP-QE/ci/blob/main/jenkins/pipelines/ci/tiflash/tiflash-build-common.groovy#L453

File: dbms/src/Common/TiFlashBuildInfo.cpp [hunk 1]
```diff
@@ -147,6 +148,14 @@ String getProfile()
     return TIFLASH_PROFILE;
 }
 
+constexpr String getCompilerVersion()
+{
+    return fmt::format(
+        "{} {}",
+        &TIFLASH_CXX_COMPILER[LogFmtDetails::getFileNameOffset(TIFLASH_CXX_COMPILER)],
```
Comment (JaySon-Huang) [inline dbms/src/Common/TiFlashBuildInfo.cpp:155]: TIFLASH_CXX_COMPILER is some strings like "/DATA/disk1/ra_common/tiflash-env-17/sysroot/bin/clang++". Use `LogFmtDetails::getFileNameOffset` to get the "clang++"

**Comment by Lloyd-Pottiger** [comment]
image

Seems it works! 🥳

**Comment by JaySon-Huang** [comment]
arm64 build job: https://ci.pingcap.net/blue/organizations/jenkins/build-common/detail/build-common/1123/pipeline/

**Review by Lloyd-Pottiger — COMMENTED** [review]
Maybe also need to update README.md?

**Comment by JaySon-Huang** [comment]
> Maybe also need to update README.md?

updated

**Comment by JaySon-Huang** [comment]
The latest MacOS 14.4.1 only provides clang 15, so I keep require AppleClang 15.0.0+ by now

**Comment by JaySon-Huang** [comment]
/hold
some ABI is changed in clang-17, so __PRETTY_FUNCTION__ is changed. Causing some ut failure

**Review by Lloyd-Pottiger — COMMENTED** [review]
image

nit: we should upgrade the ninja in the docker.

**Pull request merged**